service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/bucket/CopyGcsBucketDataStep.java

```diff
@@ -0,0 +1,270 @@
+      //   support requester pays right now for controlled gcs buckets, but it might be set on a
+      //   referenced bucket.
+      final String transferServiceSAEmail =
+          workingMap.get(ControlledResourceKeys.STORAGE_TRANSFER_SERVICE_SA_EMAIL, String.class);
+      logger.debug("Storage Transfer Service SA: {}", transferServiceSAEmail);
+
+      final TransferJob transferJobInput =
+          new TransferJob()
+              .setName(transferJobName)
+              .setDescription("Terra Workspace Manager Clone GCS Bucket")
+              .setProjectId(controlPlaneProjectId)
+              .setSchedule(createScheduleRunOnceNow())
+              .setTransferSpec(
+                  createTransferSpec(
+                      sourceInputs.getBucketName(), destinationInputs.getBucketName()))
+              .setStatus(ENABLED_STATUS);
+      // Create the TransferJob for the associated schedule and spec in the correct project.
+      final TransferJob transferJobOutput =
+          storageTransferService.transferJobs().create(transferJobInput).execute();
+      logger.debug("Created transfer job {}", transferJobOutput);
+      // Job is now submitted with its schedule. We need to poll the transfer operations API
+      // for completion of the first transfer operation. The trick is going to be setting up a
+      // polling interval that's appropriate for a wide range of bucket sizes. Everything from
+      // millisecond
+      // to hours. The transfer operation won't exist until it starts.
+      final String operationName =
+          getLatestOperationName(storageTransferService, transferJobName, controlPlaneProjectId);
+
+      final StepResult operationResult =
+          getTransferOperationResult(storageTransferService, transferJobName, operationName);
+
+      if (StepStatus.STEP_RESULT_FAILURE_FATAL == operationResult.getStepStatus()) {
+        return operationResult;
+      }
+
+      deleteTransferJob(storageTransferService, transferJobName, controlPlaneProjectId);
+    } catch (IOException e) {
+      return new StepResult(
+          StepStatus.STEP_RESULT_FAILURE_FATAL,
+          new IllegalStateException("Failed to copy bucket data", e));
+    }
+
+    final ApiClonedControlledGcpGcsBucket apiBucketResult =
+        flightContext
+            .getWorkingMap()
+            .get(
+                ControlledResourceKeys.CLONE_DEFINITION_RESULT,
+                ApiClonedControlledGcpGcsBucket.class);
+    FlightUtils.setResponse(flightContext, apiBucketResult, HttpStatus.OK);
+    return StepResult.getStepResultSuccess();
+  }
+
+  // All data in the bucket will be deleted in the previous step's undo method
+  @Override
+  public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+    return StepResult.getStepResultSuccess();
+  }
+
+  // Delete the transfer job, as we don't support reusing them.
+  private void deleteTransferJob(
```

Comment:
I don't follow the two step solution. One to do the transfer and another that just deletes the transfer job? I don't see how that helps in writing the undo code.

Reply:
This flight can leak transfers on failure or on being restarted. Is that OK?

For example, if the pod crashes at line 108, Stairway will recover the flight and run the step from the beginning. The result would be two transfers, I think. Is that the case?

Or if there is an IOException, the transfer does not get deleted. Is that a problem?

Reply:
having two transfer jobs active at once with the same sink might be OK given that we're only overwriting files that are different from the source. A given transfer job can only ever have one operation running at once, but since we create a new job every time we may be sidestepping that rule.

If we fail to delete the transfer job, there's no cost except for clutter in the console in the control plane project. There's a small risk we could inadvertently re-run the job later. That's easy to fix, though, by calling the delete method from the catch block.

Reply:
Here is my suggestion
1. Always generate the transfer job name using the flight id - not a random number. That way, rerunning steps will always use the same job name.
2. Step A: in the `do()`: looks up the transferJob. If it exists, return success. If it does not exist, create it. In the `undo()`: deletes the transferJob. Does not complain if it isn't there. So if there is a failure in any later step, the undo will delete the job.
3. Step B: wait for job completion. If the job succeeds, we return success. If it fails, we fail. Failure triggers undo and the job will be deleted. If this step is interrupted or restarts, it simply picks up where it left off. It is line 108-116, calling `getLatestOperationName` and `getTransferOperationResult` and returning the result.
4. Step C: Delete the the transferJob. No undo.
5. Finally, RemoveBucketRolesStep. No undo.

I think with that organization we do not leak transferJobs and we do not create duplicate transferJobs.

Reply:
I like it. That step was getting way too long anyway.